tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php

```diff
@@ -0,0 +1,303 @@
+                $this->_em->getClassMetadata(__NAMESPACE__ . '\DDC2363ServicesPackage'),
+                $this->_em->getClassMetadata(__NAMESPACE__ . '\DDC2363Service'),
+            )
+        );
+    }
+
+    public function testIssue()
+    {
+        // We create an Order with related ServicesPackage that is related to two Service.
+        $order = new DDC2363Order();
+        $order->setName('My ORDER');
+
+        $servicesPackage = $this->createNewServicesPackage();
+        $order->setServicesPackage($servicesPackage);
+
+        $this->_em->persist($order);
+        $this->_em->flush();
+
+        // So, we assert that there is only one ServicesPackage and 2 PurchasedService.
+        // On creation there is no problem, these assertions are green.
+        $this->assertCount(
+            1,
+            $this->_em->getRepository(__NAMESPACE__ . '\\DDC2363ServicesPackage')->findAll()
+        );
+        $this->assertCount(
+            2,
+            $this->_em->getRepository(__NAMESPACE__ . '\\DDC2363Service')->findAll()
+        );
+
+        // Reinitializing EntityManager to simulate, for example, another HTTP request.
+        $this->_em->close();
+        $this->_em = $this->_getEntityManager();
+
+        // Now we load previously persisted Order and associate it with a new ServicesPackage with two new Services.
+        $order = $this->_em->getRepository(__NAMESPACE__ . '\\DDC2363Order')->find(1);
+
+        $servicesPackage = $this->createNewServicesPackage($this->_em);
+        $order->setServicesPackage($servicesPackage);
+
+        $this->_em->persist($order);
+        $this->_em->flush();
+
+        // We assert again that there is only one ServicesPackage and two Services. This is because we have
+        // orphanRemoval on Order::purchased_services_package and cascade="remove" on
+        // PurchasedServicesPackage::services.
+        $this->assertCount(
+            1,
+            $this->_em->getRepository(__NAMESPACE__ . '\\DDC2363ServicesPackage')->findAll()
+        );
+        $this->assertCount(
+            2,
+            $this->_em->getRepository(__NAMESPACE__ . '\\DDC2363Service')->findAll()
+        );
+
+        // Reinitializing EntityManager to simulate, for example, another HTTP request.
+        $this->_em->close();
+        $this->_em = $this->_getEntityManager();
+
+        // We load again the Order...
+        $order = $this->_em->getRepository(__NAMESPACE__ . '\\DDC2363Order')->find(1);
+
+        // To make test pass, uncomment the following line. So ServicesPackage will be hydrated.
+        // $order->getServicesPackage()->getName();
+
+        // ... and create another ServicesPackage with two Services.
+        $servicesPackage = $this->createNewServicesPackage($this->_em);
+        $order->setServicesPackage($servicesPackage);
+
+        $this->_em->persist($order);
+        $this->_em->flush();
+        $this->_em->flush(); // <--- This is the problem!
```

Comment:
Something else is happening here. I went through it with the debugger, and what is happening is that the `$services_package` property is replaced by the one created at [line 60](https://github.com/doctrine/doctrine2/pull/625/files#diff-d96f5038a94052e8ae58f5a7c642fa3cR60), so this second `flush()` is working with an incorrect object graph.

The [first `flush()`](https://github.com/doctrine/doctrine2/pull/625/files#diff-d96f5038a94052e8ae58f5a7c642fa3cR93) is causing the graph to be broken

Reply:
The problem seems to be that the first `flush()` causes lazy-loading of the proxy previously referenced in the `services_package` property. This is correct, as these entities are targeted for removal. The lazy-loading is forced by `cascadeRemove()`:

https://github.com/doctrine/doctrine2/blob/374e7ace49d864dad8cddbc55346447c8a6a2083/lib/Doctrine/ORM/UnitOfWork.php#L2300-L2313

Lazy-loading of a `DDC2363ServicesPackage` object causes the owning side of the `one-to-one` association to be overwritten with a proxy reference, which means that `DDC2363Order#$services_package` now references the old lazy-loaded proxy.

This happens because the query to load the `DDC2363ServicesPackage` implicitly fetch-joins `DDC2363Order`, and hydration causes the existing `DDC2363Order#$services_package` to be replaced.

Very very very tricky, and I don't think this is fixable without completely disabling:

 * lazy-loading of cascade-removed proxies
 * fetch-joins on one-to-one associations